sql: a few compatibility features

jordanlewis opened the pull request:
- add pg_my_temp_schema() builtin
- add array overlaps (&&) operator
- add array contains (@>) and contained_by(<@) operators
- add support for SET SESSION AUTHORIZATION DEFAULT
- add pg_get_function_identity_arguments builtiin

Closes #18442.
Closes #33337.
Closes #39821.
Closes #40211.

jordanlewis committed:
builtins: add pg_my_temp_schema

Release note (sql change): add the pg_my_temp_schema builtin for
Postgres compatibility.

jordanlewis committed:
sql: add array overlaps (&&) and contains (@>, <@) ops

This commit adds support for the && operator on 2 arrays, which returns
true if the elements have at least 1 common element, and the @> and <@
operators on 2 arrays, which return true if the array on the left
contains the array or the right, or vice versa.

In addition, it refactors the && operator from being syntactic sugar for
the inet_overlaps builtin to be a normal operator.

Release note (sql change): Add the overlaps (&&), contains (@>) and
contained_by (<@) operators for arrays.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/40194)
<!-- Reviewable:end -->

jordanlewis committed:
sql: add support for SET SESSION AUTHORIZATION DEFAULT

Currently a no-op.

Release note: None

jordanlewis committed:
builtins: add pg_get_function_identity_arguments builtin

This is used by a couple of UI tools. It returns a textual list of the
argument types of a function that's identified by its OID.

Release note (sql change): add pg_get_function_identity_arguments
builtin

knz reviewed (commented):
r1 and r4 OK

There's a code mix-up between r2 and r3, so you can either fix the mix-up or squash the commits together.

r5 I really want to trust your understanding of the pg spec but there is currently no guardrail if anyone else tweaks the implementation later - so maybe more tests would help here.

> Reviewed 2 of 2 files at r1, 15 of 15 files at r2, 3 of 3 files at r3, 7 of 9 files at r4, 3 of 3 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40194#-:-LnK8LZq9I94tzisH6fh:bxts31f)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @justinj)

___
*[pkg/sql/logictest/testdata/logic_test/builtin_function, line 2252 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/40194#-LnK85UZ1ZpHNvSsRDXM:-LnK85UZ1ZpHNvSsRDXN:bnb0zzn) ([raw file](https://github.com/cockroachdb/cockroach/blob/05b1d0bce0d00c203c6e3cfb4023d4b7f372f338/pkg/sql/logictest/testdata/logic_test/builtin_function#L2252)):*
> ```
> # Sanity check pg_get_function_identity_arguments.
> query T
> SELECT pg_get_function_identity_arguments('convert_from'::regproc::oid)
> ```

I think the complexity of this function mandates a wider diversity of tests.
You probably want a mix of:
- polymorphic functions
- functions with zero arguments
- overloaded functions

and see that the behavior is consistent (and compatible with pg) for all of them.
___
*[pkg/sql/sem/builtins/pg_builtins.go, line 627 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40194#-LnK6mlu9scl0yFIqykc:-LnK6mlv7ABXRCt34QtF:b6kaqv4) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6d76ccee9480ec2fa32eb2448183c9159e668ac/pkg/sql/sem/builtins/pg_builtins.go#L627)):*
> ```Go
> 	),
> 
> 	"pg_my_temp_schema": makeBuiltin(defProps(),
> ```

Maybe add link to pg doc where this is explained in comment above.
___
*[pkg/sql/sem/tree/eval.go, line 296 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/40194#-LnK7VxM8x8d_TEXEI6K:-LnK7VxM8x8d_TEXEI6L:b-emv6mm) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebb27df162e927c4068ff30d6311e51c795324f3/pkg/sql/sem/tree/eval.go#L296)):*
> ```Go
> 
> // ArrayContains return true if the haystack contains all needles.
> func ArrayContains(haystack *DArray, needles *DArray) (*DBool, error) {
> ```

Where is this function used? It seems you got it inlined below.
___
*[pkg/sql/sem/tree/eval.go, line 303 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/40194#-LnK7AIG3WIB-Cdcv4Uv:-LnK7AIG3WIB-Cdcv4Uw:bcaz3hk) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebb27df162e927c4068ff30d6311e51c795324f3/pkg/sql/sem/tree/eval.go#L303)):*
> ```Go
> 		var found bool
> 		for j := range haystack.Array {
> 			if needles.Array[i] == haystack.Array[j] {
> ```

I think you need `Compare` here too
___
*[pkg/sql/sem/tree/eval.go, line 307 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/40194#-LnK7dlW4JYmhISilDWv:-LnK7dlW4JYmhISilDWw:by6w4ni) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cb6ebc60643bef5ed648b2a334821ae89a520d8/pkg/sql/sem/tree/eval.go#L307)):*
> ```Go
> 		var found bool
> 		for _, hay := range haystack.Array {
> 			if needle.Compare(ctx, hay) == 0 {
> ```

ok this change you can pull into the previous commit
___
*[pkg/sql/sem/tree/eval.go, line 2095 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/40194#-LnK7gg9DOKogW2h5rsa:-LnK7gg9DOKogW2h5rsb:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cb6ebc60643bef5ed648b2a334821ae89a520d8/pkg/sql/sem/tree/eval.go#L2095)):*
> ```Go
> 				haystack := MustBeDArray(left)
> 				needles := MustBeDArray(right)
> 				return ArrayContains(ctx, haystack, needles)
> ```

ditto


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
TFTR. I squashed r2 and r3 which I think resolves the issues. I added more tests as well.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40194#-:-LnKys8D5J08aiVsFs47:b2p95sj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @justinj and @knz)

___
*[pkg/sql/logictest/testdata/logic_test/builtin_function, line 2252 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/40194#-LnK85UZ1ZpHNvSsRDXM:-LnL1RTAEt2x5e7yXLJA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/05b1d0bce0d00c203c6e3cfb4023d4b7f372f338/pkg/sql/logictest/testdata/logic_test/builtin_function#L2252)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I think the complexity of this function mandates a wider diversity of tests.
You probably want a mix of:
- polymorphic functions
- functions with zero arguments
- overloaded functions

and see that the behavior is consistent (and compatible with pg) for all of them.
</blockquote></details>

Done.
___
*[pkg/sql/sem/builtins/pg_builtins.go, line 627 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40194#-LnK6mlu9scl0yFIqykc:-LnKyYN844EZCCgVaI-W:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6d76ccee9480ec2fa32eb2448183c9159e668ac/pkg/sql/sem/builtins/pg_builtins.go#L627)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Maybe add link to pg doc where this is explained in comment above.
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/eval.go, line 296 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/40194#-LnK7VxM8x8d_TEXEI6K:-LnKylIP9Mk4YlDXPldm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebb27df162e927c4068ff30d6311e51c795324f3/pkg/sql/sem/tree/eval.go#L296)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Where is this function used? It seems you got it inlined below.
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/eval.go, line 303 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/40194#-LnK7AIG3WIB-Cdcv4Uv:-LnKykdxCweQfMGL7B3D:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebb27df162e927c4068ff30d6311e51c795324f3/pkg/sql/sem/tree/eval.go#L303)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I think you need `Compare` here too
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/eval.go, line 307 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/40194#-LnK7dlW4JYmhISilDWv:-LnKyktN3FqeYuLE6QcP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cb6ebc60643bef5ed648b2a334821ae89a520d8/pkg/sql/sem/tree/eval.go#L307)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

ok this change you can pull into the previous commit
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/eval.go, line 2095 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/40194#-LnK7gg9DOKogW2h5rsa:-LnKymSVCGC9S4AlxSME:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cb6ebc60643bef5ed648b2a334821ae89a520d8/pkg/sql/sem/tree/eval.go#L2095)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
LGTM with two nits

> Reviewed 13 of 15 files at r7, 9 of 9 files at r8, 3 of 3 files at r9.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40194#-:-LnMDum70nA-WLg0EAU6:b-n5dmbj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @justinj)

___
*[pkg/sql/parser/sql.y, line 3126 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/40194#-LnMDMxlCRTehoqUG2s7:-LnMDMxlCRTehoqUG2s8:b-2ft1p4) ([raw file](https://github.com/cockroachdb/cockroach/blob/553689136b944bab33e2c57925d3dab7518ece53/pkg/sql/parser/sql.y#L3126)):*
> ```Yacc
> | SESSION AUTHORIZATION non_reserved_word_or_sconst
>   {
>     return unimplemented(sqllex, "set session authorization name")
> ```

maybe file an issue and link it here? I understand that there are still several errors without issue number throughout the grammar, but I think Andy was asking that any new such errors should be paired with an issue.
___
*[pkg/sql/sem/builtins/pg_builtins.go, line 615 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/40194#-LnMDcUJ0YUgcfavVAf8:-LnMDcUJ0YUgcfavVAf9:boj49a3) ([raw file](https://github.com/cockroachdb/cockroach/blob/553689136b944bab33e2c57925d3dab7518ece53/pkg/sql/sem/builtins/pg_builtins.go#L615)):*
> ```Go
> 	// pg_get_function_identity_arguments returns the argument list necessary to
> 	// identify a function, in the form it would need to appear in within ALTER
> 	// FUNCTION, for instance. This form omits default values.
> ```

nit: URL to pg doc here too


<!-- Sent from Reviewable.io -->
